### PR TITLE
Modification to make PHP 7.2 compatible

### PR DESCRIPTION
**Pull request by olecorre**
1.6.1.25 (2019-08-26)

classes/AdminTab.php line 511 change "$function($value)"

classes/AdminTab.php line 530 change "$function($value)"

classes/AdminTab.php line 1675 change "create_function"

classes/Cart.php line 2772 add control "is_array"

classes/Cookie.php line 79 change "else" and add 2 constants OPENSSL_KEY and OPENSSL_IV
Mcrypt disappears. I had Warning issues with existing keys that are not good for Openssl. So I added two new constants with the correctly generated keys.

classes/ImageType.php line 173 change "if"

classes/ObjectModel.php line 64 change "type of variable"

classes/Rijndael.php line 97 delete constant "MCRYPT_RIJNDAEL_128" it's not exists

classes/Tools.php line 2151 change "create_function"

classes/Tools.php line 3736 change "create_function"

classes/controller/AdminController.php line 549 change "paramaters" for function "initBreadcrumbs"

classes/controller/AdminController.php line 2329 change "if"

classes/controller/AdminController.php line 3467 change "$function($value)"

classes/controller/AdminController.php line 3469 change "$function($value)"

classes/helper/HelperList.php line 198 change "create_function"

classes/module/Module.php line 1554 change "create_function"

classes/stock/SupplyOrderDetail.php line 304 change "$function($value)"

controllers/admin/AdminAttributeGeneratorController.php line 79 change "create_function"

controllers/admin/AdminOrdersController.php line 589 change "$function"

controllers/admin/AdminProductsController.php line 2221 change "$function"

controllers/admin/AdminProductsController.php line 2224 change "$function"

controllers/admin/AdminProductsController.php line 2240 change "$function"

controllers/front/ProductController.php line 254 add controle "is_array"

install-dev/install_version.php line 27 change version

install-dev/fixtures/fashion/data/generate_attribute.php line 203 change "create_function"

install-dev/upgrade/upgrade.php line 258 add constants OPENSSL_KEY and OPENSSL_IV

install-dev/models/install.php line 93 add create constants OPENSSL_KEY and OPENSSL_IV

tools/profiling/Controller.php line 358 change "create_function"

tools/smarty/sysplugins/smarty_internet_compilebase.php rewrite function "getAttributes" for delete function "each"
the smarty update was too much change, so I decided to redo one of the functions so that it would be compatible. I had no mistakes after this correction

tools/tcpdf update TCPDF to the the last version 6.2.26

admin-dev/themes/default/template/controllers/modules/js.tpl line 37 change "count" to "is_array"

admin-dev/themes/default/template/controllers/information/helpers/view/view.tpl

### CODE REVIEW

**Review by Max84 — DISMISSED** [review]
Great job !

**Comment by sunereyt** [comment]
That sounds very good !
Looking forward 1.6.1.24 :)
Thanks Olecorre

**Comment by jeckyl** [comment]
I hope that this version will soon be officially pushed to help merchants with the changes of unscrupulous hosts and official partners of PrestaShop.

**Comment by olecorre** [comment]
On a manual correction of 1.6.1.24 for PHP 7.3, it is necessary in a separate file to make an echo base64_encode(openssl_random_pseudo_bytes(openssl_cipher_iv_length('AES-128-CBC')));
and add two lines in the settings.inc.php for the two constants _OPENSSL_KEY_ and _OPENSSL_IV_

**Comment by ghost** [comment]
Work well on 
Plesk 17.5
PHP : 7.2.26 

if you need more informations tell me

**Comment by ttoine** [comment]
Needs another reviewer from one of the 4 volunteers to merge it.
@kpodemski could you please test and confirm on your side, and then approve it if everything is ok?

**Comment by ghost** [comment]
It's work for me same config :

Plesk 17.5
PHP : 7.2.26

**Review by jeckyl — APPROVED** [review]
I applied as is all the modifications for a PHP7.3 upgrade and I didn't encounter any problems other than third party modules.

**Review by kpodemski — DISMISSED** [review]

**Comment by kpodemski** [comment]
Good job 👍

**Comment by ghost** [comment]
@jeckyl && @Max84 we need one more review :)

**Review by jeckyl — APPROVED** [review]
I approve

**Comment by jeckyl** [comment]
> @jeckyl && @Max84 we need one more review :)

As you can see, I have twice validated the changes but I don't have a status allowing me to do so.
Capture d'écran 2020-03-05 20 38 37

**Comment by kevin39** [comment]
Working fine here. Just upgraded an old 1.5 installation on debian 6 to last 1.6 on debian 8+PHP 7.2 and everything is working fine with these fixes (and with a little luck, no third party modules crashed for me).

**Comment by MildConcussion** [comment]
> On a manual correction of 1.6.1.24 for PHP 7.3, it is necessary in a separate file to make an echo base64_encode(openssl_random_pseudo_bytes(openssl_cipher_iv_length('AES-128-CBC')));
> and add two lines in the settings.inc.php for the two constants _OPENSSL_KEY_ and _OPENSSL_IV_

In this case you get one value, which looks like IV. Does the key remain same or?

**Comment by olecorre** [comment]
the line of code is for the value of _OPENSSL_IV_. For OPENSSL_KEY I copy the value of _COOKIE_IV_

**Comment by ghost** [comment]
Great job @olecorre . We successfully tested it on multiple occasions. It's as well compatible with PHP 7.3.
"up" for merge approval :)

**Comment by kevin39** [comment]
Confirmed here too. In production since few months, no issue.
Good job !

**Comment by MildConcussion** [comment]
Can confirm as well. In production for a few months, people are buying with no problems :)

**Review by jeckyl — DISMISSED** [review]

**Review by Max84 — DISMISSED** [review]
I didn't see last ping, sorry.
Update is ok, but you should fix automated test (check "Undefined index: $column" with Travis).

File: classes/Rijndael.php
Comment (s7anley) [inline classes/Rijndael.php:131]: I think you need to rather hard-code name of the cipher otherwise old cookies will not work, see also https://github.com/PrestaShop/PrestaShop/blob/42833c4097071b46f37b76f6c347fb579447e3e1/classes/PhpEncryptionLegacyEngine.php#L135.

**Comment by devth8** [comment]
Why did you make this type of modification:
classes/AdminTab.php line 511 change "$function($value)"
By "::{$function}" ?

While there are countless accesses of the same type:
::$

Example:
if (!Validate::$values_validation(Tools::getValue($field.'_'.$id_lang))) {
line 1020 of the same file
?

Thanks

**Comment by mathouse88** [comment]
> Can confirm as well. In production for a few months, people are buying with no problems :)

Noob question: you made changes manually or there is an auto-update solution ?

**Comment by MildConcussion** [comment]
> > Can confirm as well. In production for a few months, people are buying with no problems :)
> 
> Noob question: you made changes manually or there is an auto-update solution ?

I made changes manually. Not aware of any auto-update one and didn't want to replace everything, just in case.

**Comment by mathouse88** [comment]
> > > Can confirm as well. In production for a few months, people are buying with no problems :)
> > 
> > 
> > Noob question: you made changes manually or there is an auto-update solution ?
> 
> I made changes manually. Not aware of any auto-update one and didn't want to replace everything, just in case.

Thank you

**Comment by ghost** [comment]
I upgraded Travis configuration to make the build compatible with PHP 7.0, 7.1, 7.2 and 7.3.
Support for PHP 5.4 and 5.5 has been removed due to compatibility issues with `travis-scripts/install-prestashop.sh`.

**Review by jeckyl — APPROVED** [review]
Everything is approved since the last modifications are to satisfy Travis.

**Review by kpodemski — APPROVED** [review]

**Comment by ttoine** [comment]
Should a beta version be available first, so it's possible to test that it works well with auto upgrade?
If a build and release is distributed via GitHub, a looot of shops will want to update. So the more it's tested, the more it will be reliable for non tech people.

**Comment by Max84** [comment]
Hi Folks,
Tested today here and read commit, I didn't see error.

**Review by Max84 — APPROVED** [review]
Good job, congratulations!

**Comment by Max84** [comment]
@ttoine It's merged. What is the next step in order to release beta version?

**Comment by ttoine** [comment]
I you feel confident about it, please just ask that a release to be built. And a dev will do it.
But not on a Friday evening. So let's start that on Monday.

BTW, quick reminder as it will be the first time such a release will be done, please, everyone involved, be tolerant and patient. The process can then be documented so it become easy.

**Comment by Bondatyandco** [comment]
Thank's
Work well prestashop 1.6.1.4 
Step 1 php 7.0 with modifications for php 7 some time ago 
Step 2 php 7.3.28 with current modifications 
happy not to upgrade to prestashop 1.7

**Comment by ElCapitan63** [comment]
I'm running your code on PHP 7.3.33 and ran into problem when trying to save an SQL in Prestashop SQL Administrator. Both issue and solution is described here (https://stackoverflow.com/questions/50148055/declaration-of-replaceprocessorprocesstokenlist-must-be-compatible-with-ins). I've applied the solution for both ReplaceProcessor.php and DuplicateProcessor.php in /tools/parser_sql/processors/

**Pull request merged**
